src/components/in-world-hud.js

```diff
@@ -0,0 +1,155 @@
+
+    const muted = this.el.sceneEl.is("muted");
+    this.mic.setAttribute("src", muted ? "#muted" : "#unmuted");
+
+    const avatarScale = "0.1 0.1 0.1";
+    const flipXAvatarScale = "-" + avatarScale;
+
+    const scene = this.el.sceneEl;
+    this.onUsernameChanged = this.onUsernameChanged.bind(this);
+    scene.addEventListener("username-changed", this.onUsernameChanged);
+
+    this.onNametagHovered = () => {
+      this.nametag.setAttribute("color", "#00eeee");
+      this.data.haptic.emit("haptic_pulse", { intensity: "low" });
+    };
+    this.onNametagUnhovered = () => {
+      this.nametag.setAttribute("color", "white");
+    };
+    this.onAvatarHovered = () => {
+      this.avatar.setAttribute("scale", flipXAvatarScale);
+      this.data.haptic.emit("haptic_pulse", { intensity: "low" });
+    };
+    this.onAvatarUnhovered = () => {
+      this.avatar.setAttribute("scale", avatarScale);
+    };
+    this.onMicHover = () => {
+      this.hoveredOnMic = true;
+      const muted = scene.is("muted");
+      this.mic.setAttribute("src", muted ? "#unmuted" : "#muted");
+      this.data.haptic.emit("haptic_pulse", { intensity: "low" });
+    };
+
+    this.showCorrectMuteState = () => {
+      const muted = this.el.sceneEl.is("muted");
+      this.mic.setAttribute("src", muted ? "#muted" : "#unmuted");
+    };
+
+    this.onStateChange = evt => {
+      if (evt.detail !== "muted") return;
+      this.showCorrectMuteState();
+    };
+
+    this.onMicHoverExit = () => {
+      this.hoveredOnMic = false;
+      this.showCorrectMuteState();
+    };
+
+    this.onSelect = evt => {
+      if (this.hoveredOnMic) {
+        this.el.emit("action_mute");
+        this.data.haptic.emit("haptic_pulse", { intensity: "low" });
+      }
+    };
+
+    this.onClick = () => {
+      this.el.emit("action_select_hud_item");
+    };
+
+    this.onAudioFrequencyChange = e => {
+      const red = 1.0 - e.detail.volume / 10.0;
+      this.mic.object3DMap.mesh.material.color = { r: red, g: 9, b: 9 };
+    };
+    this.el.sceneEl.addEventListener("mediaStream", evt => {
+      this.ms = evt.detail.ms;
+      const ctx = THREE.AudioContext.getContext();
+      const source = ctx.createMediaStreamSource(this.ms);
+      this.analyser = ctx.createAnalyser();
+      this.levels = new Uint8Array(this.analyser.frequencyBinCount);
+      source.connect(this.analyser);
+    });
+  },
+
+  play() {
+    this.mic.addEventListener("raycaster-intersected", this.onMicHover);
+    this.mic.addEventListener("raycaster-intersected-cleared", this.onMicHoverExit);
+
+    this.nametag.addEventListener("raycaster-intersected", this.onNametagHovered);
+    this.nametag.addEventListener("raycaster-intersected-cleared", this.onNametagUnhovered);
+
+    this.avatar.addEventListener("raycaster-intersected", this.onAvatarHovered);
+    this.avatar.addEventListener("raycaster-intersected-cleared", this.onAvatarUnhovered);
```

Comment:
Seems like we should create a button element/component that encapsulates haptic and hover states.

Reply:
Yep, I avoided suggesting that because I am considering the entire 2d-hud and in-world-hud to be throw away.